src/main/java/io/pinecone/PineconeDataPlaneClient.java

```diff
@@ -0,0 +1,222 @@
+package io.pinecone;
+
+import com.google.protobuf.Struct;
+import io.pinecone.exceptions.PineconeValidationException;
+import io.pinecone.proto.*;
+
+import java.util.ArrayList;
+import java.util.List;
+
+public class PineconeDataPlaneClient {
+    private VectorServiceGrpc.VectorServiceBlockingStub blockingStub;
+
+    public PineconeDataPlaneClient(VectorServiceGrpc.VectorServiceBlockingStub blockingStub) {
+        if (blockingStub == null) {
+            throw new PineconeValidationException("BlockingStub cannot be null.");
+        }
+
+        this.blockingStub = blockingStub;
+    }
+
+    public UpsertResponse upsert(String id,
+                                 List<Float> values) {
+        return upsert(id, values, null, null, null, null);
+    }
+
+    public UpsertResponse upsert(String id,
+                                 List<Float> values,
+                                 String namespace) {
+        return upsert(id, values, null, null, null, namespace);
+    }
+
+    public UpsertResponse upsert(String id,
+                                 List<Float> values,
+                                 List<Long> sparseIndices,
+                                 List<Float> sparseValues,
+                                 com.google.protobuf.Struct metadata,
+                                 String namespace) {
+        UpsertRequest.Builder upsertRequest = UpsertRequest.newBuilder();
+
+        if (id == null || id.isEmpty() || values == null || values.isEmpty()) {
+            throw new PineconeValidationException("Invalid Upsert Request. Please ensure that both id and values are provided.");
+        }
+
+        Vector.Builder vectorBuilder = Vector.newBuilder()
+                .setId(id)
+                .addAllValues(values);
+
+        if (sparseIndices != null && sparseValues != null && sparseIndices.size() != sparseValues.size()) {
+            throw new PineconeValidationException("Invalid Upsert Request. Please ensure that both sparse indices and values are of the same length.");
+        }
+
+        if (sparseIndices != null && !sparseIndices.isEmpty()) {
+            vectorBuilder.setSparseValues(SparseValues.newBuilder()
+                    .addAllIndices(convertUnsigned32IntToSigned32Int(sparseIndices))
+                    .addAllValues(sparseValues)
+                    .build());
+        }
+
+        if (metadata != null) {
+            vectorBuilder.setMetadata(metadata);
+        }
+
+        upsertRequest.addVectors(vectorBuilder.build());
+
+        if (namespace != null && !namespace.isEmpty()) {
+            upsertRequest.setNamespace(namespace);
+        }
+
+        return blockingStub.upsert(upsertRequest.build());
+    }
+
+    public QueryResponse query(String id,
```

Comment:
I haven't added a separate query() since topk is the only required parameter and passing only top k while setting everything else to null is not going to be a widely used method